dialog/color_wheel.go

```diff
@@ -0,0 +1,209 @@
+	}
+	a.ExtendBaseWidget(a)
+	return a
+}
+
+// Cursor returns the cursor type of this widget.
+func (a *colorWheel) Cursor() desktop.Cursor {
+	return desktop.CrosshairCursor
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer.
+func (a *colorWheel) CreateRenderer() fyne.WidgetRenderer {
+	raster := &canvas.Raster{
+		Generator: a.generator,
+	}
+	x := canvas.NewLine(color.Black)
+	y := canvas.NewLine(color.Black)
+	return &colorWheelRenderer{
+		BaseRenderer: internalwidget.NewBaseRenderer([]fyne.CanvasObject{raster, x, y}),
+		area:         a,
+		raster:       raster,
+		x:            x,
+		y:            y,
+	}
+}
+
+// MinSize returns the size that this widget should not shrink below.
+func (a *colorWheel) MinSize() fyne.Size {
+	a.ExtendBaseWidget(a)
+	return a.BaseWidget.MinSize()
+}
+
+// SetHSLA updates the selected color in the wheel.
+func (a *colorWheel) SetHSLA(hue, saturation, lightness, alpha int) {
+	if a.Hue == hue && a.Saturation == saturation && a.Lightness == lightness && a.Alpha == alpha {
+		return
+	}
+	a.Hue = hue
+	a.Saturation = saturation
+	a.Lightness = lightness
+	a.Alpha = alpha
+	a.Refresh()
+}
+
+// Tapped is called when a pointer tapped event is captured and triggers any change handler.
+func (a *colorWheel) Tapped(event *fyne.PointEvent) {
+	a.trigger(event.Position)
+}
+
+// Dragged is called when a pointer drag event is captured and triggers any change handler
+func (a *colorWheel) Dragged(event *fyne.DragEvent) {
+	a.trigger(event.Position)
+}
+
+// DragEnd is called when a pointer drag ends
+func (a *colorWheel) DragEnd() {
+}
+
+func (a *colorWheel) colorAt(x, y, w, h int) color.Color {
+	width, height := float64(w), float64(h)
+	dx := float64(x) - (width / 2.0)
+	dy := float64(y) - (height / 2.0)
+	radius, radians := cmplx.Polar(complex(dx, dy))
+	limit := math.Min(width, height) / 2.0
+	if radius > limit {
+		// Out of bounds
+		return theme.BackgroundColor()
+	}
+	degrees := radians * (180.0 / math.Pi)
+	hue := wrapHue(int(degrees))
+	saturation := int(radius / limit * 100.0)
+	red, green, blue := hslToRgb(hue, saturation, a.Lightness)
+	return &color.NRGBA{
+		R: uint8(red),
+		G: uint8(green),
+		B: uint8(blue),
+		A: uint8(a.Alpha),
+	}
+}
+
+func (a *colorWheel) locationForPosition(pos fyne.Position) (x, y int) {
+	can := fyne.CurrentApp().Driver().CanvasForObject(a)
+	x, y = pos.X, pos.Y
+	if can != nil {
+		x, y = can.PixelCoordinateForPosition(pos)
+	}
+	return
+}
+
+func (a *colorWheel) selection(width, height int) (int, int) {
+	w, h := float64(width), float64(height)
+	radius := float64(a.Saturation) / 100.0 * math.Min(w, h) / 2.0
+	degrees := float64(a.Hue)
+	radians := degrees * math.Pi / 180.0
+	c := cmplx.Rect(radius, radians)
+	return int(real(c) + w/2.0), int(imag(c) + h/2.0)
+}
+
+func (a *colorWheel) trigger(pos fyne.Position) {
+	x, y := a.locationForPosition(pos)
+	if c := a.cache; c != nil {
+		b := c.Bounds()
+		if f := a.onChange; f != nil {
```

Comment:
The body of this function is nested in two nil checks, can it be cleaner with corner cases exit early?
The a.Refresh could be in defer if it needs to run last no matter what

Reply:
Fixed! Tho I merged the `if`s rather than returning early